crates/polars-plan/src/logical_plan/optimizer/cluster_with_columns.rs

```diff
@@ -0,0 +1,248 @@
+
+            // @NOTE
+            // We can pushdown any column that utilizes no live columns that are generated in the
+            // input.
+
+            let mut column_map = ColumnMap::default();
+
+            let input_gen_columns: Vec<usize> = input_wc
+                .exprs
+                .as_exprs()
+                .iter()
+                .filter_map(|expr| match expr.output_name_inner() {
+                    OutputName::None => None,
+                    OutputName::LiteralLhs(s) | OutputName::ColumnLhs(s) => {
+                        Some(column_map_idx(&mut column_map, s.clone()))
+                    },
+                    OutputName::Alias(s) => Some(column_map_idx(&mut column_map, s.clone())),
+                })
+                .collect();
+
+            let current_livesets: Vec<LiveSet> = current_wc
+                .exprs
+                .as_exprs()
+                .iter()
+                .map(|expr| LiveSet::new(expr.node(), expr_arena, &mut column_map))
+                .collect();
+            let current_livesets: Vec<Bitmap> = current_livesets
+                .into_iter()
+                .map(|s| s.freeze(&column_map))
+                .collect();
+
+            // @NOTE: This satisfies the borrow checker
+            let num_names = column_map.len();
+            drop(column_map);
+
+            let mut input_generate = MutableBitmap::from_len_zeroed(num_names);
+            for idx in &input_gen_columns {
+                input_generate.set(*idx, true);
+            }
+            let input_generate = input_generate.freeze();
+
+            // Check for every expression in the current WITH_COLUMNS node whether it can be pushed
+            // down.
+            let mut pushable = MutableBitmap::with_capacity(input_wc.exprs.len());
+            for expr_liveset in &current_livesets {
+                let has_intersection = input_generate.intersect_with(expr_liveset);
+                let is_pushable = !has_intersection;
+                pushable.push(is_pushable);
+            }
+            let pushable = pushable.freeze();
+
+            // There is nothing to push down. Move on.
+            if pushable.set_bits() == 0 {
+                break;
+            }
+
+            // If all columns are pushable, we can merge the input into the current. This should be
+            // a relatively common case.
+            if pushable.set_bits() == pushable.len() {
+                // @NOTE: To keep the schema correct, we reverse the order here. As a
+                // `WITH_COLUMNS` higher up produces later columns. This also allows us not to
+                // have to deal with schemas.
+                input_wc
+                    .exprs
+                    .exprs_mut()
+                    .extend(std::mem::take(current_wc.exprs.exprs_mut()));
+                std::mem::swap(current_wc.exprs.exprs_mut(), input_wc.exprs.exprs_mut());
+
+                *current_wc.input = *input_wc.input;
```

Comment:
Nice trick. I was looking where you overwrote `current` so that you continue traversing the multiple `wc`'s, but you change the input.

I think we do need a comment here explaining that this drives traversal of the tree.